data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java

```diff
@@ -118,18 +118,34 @@ private AcknowledgementSet createAcknowledgementSet(Map<TopicPartition, Range<Lo
                 if (result == true) {
                     positiveAcknowledgementSetCounter.increment();
                     offsets.forEach((partition, offsetRange) -> {
-                        int partitionId = partition.partition();
-                        if (!partitionCommitTrackerMap.containsKey(partitionId)) {
-                            OffsetAndMetadata committedOffsetAndMetadata = consumer.committed(partition);
-                            Long committedOffset = Objects.nonNull(committedOffsetAndMetadata) ? committedOffsetAndMetadata.offset() : null;
-
-                            partitionCommitTrackerMap.put(partitionId, new TopicPartitionCommitTracker(partition, committedOffset));
+                        try {
+                            int partitionId = partition.partition();
+                            if (!partitionCommitTrackerMap.containsKey(partitionId)) {
+                                OffsetAndMetadata committedOffsetAndMetadata = null;
+                                synchronized(consumer) {
+                                    committedOffsetAndMetadata = consumer.committed(partition);
```

Comment:
Is Kafka consumer not thread safe? Does this not just ask the server to commit?

Reply:
Kafka consumer is not thread-safe. Any usage to "consumer" must be protected across different threads. I was getting concurrent usage error.

Reply:
https://www.confluent.io/blog/kafka-consumer-multi-threaded-messaging/

if we can avoid synchronization to commit offsets. I didnt see this blog using synchronization.

Reply:
That post explicitly states:

> When implementing a multi-threaded consumer architecture, it is important to note that the Kafka consumer is not thread safe. Multi-threaded access must be properly synchronized, which can be tricky. This is why the single-threaded model is commonly used.